internal/versions/versions.go

```diff
@@ -115,21 +118,34 @@ func (h *handler) GetReleaseImage(ctx context.Context, openshiftVersion, cpuArch
 	if err := h.kubeClient.List(ctx, clusterImageSets); err != nil {
 		return nil, err
 	}
+	var wg sync.WaitGroup
 	for _, clusterImageSet := range clusterImageSets.Items {
-		existsInCache := false
-		for _, releaseImage := range h.releaseImages {
-			if releaseImage.URL != nil && *releaseImage.URL == clusterImageSet.Spec.ReleaseImage {
-				existsInCache = true
-				break
-			}
+		if err := h.sem.Acquire(ctx, 1); err != nil {
+			// don't fail the entire function if this iteration fails to acquire the semaphore
+			continue
 		}
-		if !existsInCache {
-			_, err := h.addReleaseImage(clusterImageSet.Spec.ReleaseImage, pullSecret)
-			if err != nil {
-				h.log.WithError(err).Warnf("Failed to add release image %s", clusterImageSet.Spec.ReleaseImage)
+		wg.Add(1)
+		go func(clusterImageSet hivev1.ClusterImageSet) {
+			defer func() {
+				wg.Done()
+				h.sem.Release(1)
+			}()
+			existsInCache := false
+			for _, releaseImage := range h.releaseImages {
+				if releaseImage.URL != nil && *releaseImage.URL == clusterImageSet.Spec.ReleaseImage {
+					existsInCache = true
+					break
+				}
 			}
-		}
+			if !existsInCache {
+				_, err := h.addReleaseImage(clusterImageSet.Spec.ReleaseImage, pullSecret)
```

Comment:
Not sure we always validate that the pullSecret exists before calling GetReleaseImage.
So maybe worth an early return before starting the cache in case the pullSecret is missing?
I.e. so we won't try to cache all images and fail on missing pullSecret.

Reply:
Again, this is just the same logic, but indented to be in a go routine. I'd prefer not to change it here unless running these in parallel somehow changes something about the pull secret.